metrics/micrometer/src/main/java/com/datastax/oss/driver/internal/metrics/micrometer/MicrometerNodeMetricUpdater.java

```diff
@@ -31,13 +31,18 @@
 import io.micrometer.core.instrument.Timer;
 import java.time.Duration;
 import java.util.Set;
+import java.util.function.Supplier;
 import net.jcip.annotations.ThreadSafe;
 
 @ThreadSafe
 public class MicrometerNodeMetricUpdater extends MicrometerMetricUpdater<NodeMetric>
     implements NodeMetricUpdater {
 
   private final Node node;
+  private final Supplier<Number> openConnectionsSupplier;
```

Comment:
Why create fields for these? They are not referenced anywhere. The original class was fine imho.

Reply:
This is really the heart of the fix.  In the original implementation the supplier was passed inline to Micrometer; there was no ref kept anywhere else.  The unexpected (at least to me) [strong reference used in the Gauge impl](https://github.com/micrometer-metrics/micrometer/blob/v1.6.5/micrometer-core/src/main/java/io/micrometer/core/instrument/Gauge.java#L46-L61) was the only thing preventing the Supplier from getting GC'd.  And that was entirely a side-effect of the constructor we were using.  So if we then change the Gauge implementation to not using strong references the only ref for the Supplier is now the weak ref in the Gauge... which means it's eligible for GC earlier than we'd like.

Creating a strong ref on the class avoids that issue; the key question is where.  It seemed reasonable to define it on the updater itself so that it won't persist beyond the lifetime of this class.

I very much grant you this is all very indirect and non-intuitive.  As a result the impl is brittle and (at a bare minimum) needs some additional commentary to explain what's going on.

Reply:
Oooh I see, we are keeping a strong ref here to compensate for the weak ref there. Smart :-)

But I agree with you that we need to better document the intent. Not only IDEs will complain (mine is complaining), but it won't be long until someone refactors this and removes the final fields, thus silently re-introducing the memory leak.

A simple trick to calm down IDEs is to move all the calls to `initializeGauge` to a separate method:

```java
  public MicrometerNodeMetricUpdater(
      Node node,
      InternalDriverContext context,
      Set<NodeMetric> enabledMetrics,
      MeterRegistry registry) {
    super(context, enabledMetrics, registry);
    this.node = node;
    // Keep a strong reference to all gauge suppliers, otherwise they will be GC'ed too soon
    openConnectionsSupplier = node::getOpenConnections;
    streamSupplier = () -> availableStreamIds(node);
    inFlightRequestsSupplier = () -> inFlightRequests(node);
    orphanedStreamIdsSupplier = () -> orphanedStreamIds(node);
    initMetrics();
  }

  private void initMetrics() {
    DriverExecutionProfile profile = context.getConfig().getDefaultProfile();
    initializeGauge(DefaultNodeMetric.OPEN_CONNECTIONS, profile, openConnectionsSupplier);
    initializeGauge(DefaultNodeMetric.AVAILABLE_STREAMS, profile, streamSupplier);
    initializeGauge(DefaultNodeMetric.IN_FLIGHT, profile, inFlightRequestsSupplier);
    initializeGauge(DefaultNodeMetric.ORPHANED_STREAMS, profile, orphanedStreamIdsSupplier);
    // ...
```